In [1]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
data=pd.read_csv('/Users/rishismac/Downloads/Spam (1).csv')
data.head()


word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_hash  \
0          0.0        0.778        0.000           0.000   
1          0.0        0.372        0.180           0.048   
2          0.0        0.276        0.184           0.010   
3          0.0        0.137        0.000           0.000   
4          0.0        0.135        0.000           0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4601 entries, 0 to 4600
Data columns (total 58 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   word_freq_make              4601 non-null   float64
 1   word_freq_address           4601 non-null   float64
 2   word_freq_all               4601 non-null   float64
 3   word_freq_3d                4601 non-null   float64
 4   word_freq_our               4601 non-null   float64
 5   word_freq_over              4601 non-null   float64
 6   word_freq_remove            4601 non-null   float64
 7   word_freq_internet          4601 non-null   float64
 8   word_freq_order             4601 non-null   float64
 9   word_freq_mail              4601 non-null   float64
 10  word_freq_receive           4601 non-null   float64
 11  word_freq_will              4601 non-null   float64
 12  word_freq_people            4601 non-null   float64
 13  word_freq_report            4601 

In [4]:
data.describe()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
count     4601.000000        4601.000000    4601.000000   4601.000000   
mean         0.104553           0.213015       0.280656      0.065425   
std          0.305358           1.290575       0.504143      1.395151   
min          0.000000           0.000000       0.000000      0.000000   
25%          0.000000           0.000000       0.000000      0.000000   
50%          0.000000           0.000000       0.000000      0.000000   
75%          0.000000           0.000000       0.420000      0.000000   
max          4.540000          14.280000       5.100000     42.810000   

       word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
count    4601.000000     4601.000000       4601.000000         4601.000000   
mean        0.312223        0.095901          0.114208            0.105295   
std         0.672513        0.273824          0.391441            0.401071   
min         0.000000        0.000000          0.000000            0.000000   
25%         0.000000        0.000000          0.000000            0.000000   
50%         0.000000        0.000000          0.000000            0.000000   
75%         0.380000        0.000000          0.000000            0.000000   
max        10.000000        5.880000          7.270000           11.110000   

       word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
count      4601.000000     4601.000000  ...  4601.000000  4601.000000   
mean          0.090067        0.239413  ...     0.038575     0.139030   
std           0.278616        0.644755  ...     0.243471     0.270355   
min           0.000000        0.000000  ...     0.000000     0.000000   
25%           0.000000        0.000000  ...     0.000000     0.000000   
50%           0.000000        0.000000  ...     0.000000     0.065000   
75%           0.000000        0.160000  ...     0.000000     0.188000   
max           5.260000       18.180000  ...     4.385000     9.752000   

       char_freq_[  char_freq_!  char_freq_$  char_freq_hash  \
count  4601.000000  4601.000000  4601.000000     4601.000000   
mean      0.016976     0.269071     0.075811        0.044238   
std       0.109394     0.815672     0.245882        0.429342   
min       0.000000     0.000000     0.000000        0.000000   
25%       0.000000     0.000000     0.000000        0.000000   
50%       0.000000     0.000000     0.000000        0.000000   
75%       0.000000     0.315000     0.052000        0.000000   
max       4.081000    32.478000     6.003000       19.829000   

       capital_run_length_average  capital_run_length_longest  \
count                 4601.000000                 4601.000000   
mean                     5.191515                   52.172789   
std                     31.729449                  194.891310   
min                      1.000000                    1.000000   
25%                      1.588000                    6.000000   
50%                      2.276000                   15.000000   
75%                      3.706000                   43.000000   
max                   1102.500000                 9989.000000   

       capital_run_length_total         spam  
count               4601.000000  4601.000000  
mean                 283.289285     0.394045  
std                  606.347851     0.488698  
min                    1.000000     0.000000  
25%                   35.000000     0.000000  
50%                   95.000000     0.000000  
75%                  266.000000     1.000000  
max                15841.000000     1.000000  

[8 rows x 58 columns]

In [5]:
#checking the spam percentage for balance of data:
data['spam'].sum()/len(data)*100

39.404477287546186

### With 39.4 percent of spam emails in the total dataset,it is evident that the dataset is balanced

In [6]:
y=np.array(data.pop('spam')).reshape(-1,1)
X=data
from sklearn.preprocessing import MinMaxScaler
scaler=MinMaxScaler()
X_scaled=scaler.fit_transform(X)


In [7]:
help(scaler)

Help on MinMaxScaler in module sklearn.preprocessing._data object:

class MinMaxScaler(sklearn.base.OneToOneFeatureMixin, sklearn.base.TransformerMixin, sklearn.base.BaseEstimator)
 |  MinMaxScaler(feature_range=(0, 1), *, copy=True, clip=False)
 |  
 |  Transform features by scaling each feature to a given range.
 |  
 |  This estimator scales and translates each feature individually such
 |  that it is in the given range on the training set, e.g. between
 |  zero and one.
 |  
 |  The transformation is given by::
 |  
 |      X_std = (X - X.min(axis=0)) / (X.max(axis=0) - X.min(axis=0))
 |      X_scaled = X_std * (max - min) + min
 |  
 |  where min, max = feature_range.
 |  
 |  This transformation is often used as an alternative to zero mean,
 |  unit variance scaling.
 |  
 |  Read more in the :ref:`User Guide <preprocessing_scaler>`.
 |  
 |  Parameters
 |  ----------
 |  feature_range : tuple (min, max), default=(0, 1)
 |      Desired range of transformed data.
 |  
 |  copy : b

In [8]:
#Importing the ML libraries and packages:
from sklearn.model_selection import train_test_split,KFold,GridSearchCV
from sklearn.svm import SVC
from sklearn.metrics import confusion_matrix


In [9]:
#Splitting the data to train and test set:
X_train,X_test,y_train,y_test=train_test_split(X,y,train_size=0.7,random_state=100)
X_train.shape,X_test.shape,y_train.shape,y_test.shape

((3220, 57), (1381, 57), (3220, 1), (1381, 1))

In [10]:
#Checking the balance of spam mails in both training and testing data:
print('train spam percentage:',y_train.sum()/len(y_train)*100)
print('test spam percentage:',y_test.sum()/len(y_test)*100)

train spam percentage: 38.60248447204969
test spam percentage: 41.27443881245475


In [11]:
help(GridSearchCV)
help(svc)

Help on class GridSearchCV in module sklearn.model_selection._search:

class GridSearchCV(BaseSearchCV)
 |  GridSearchCV(estimator, param_grid, *, scoring=None, n_jobs=None, refit=True, cv=None, verbose=0, pre_dispatch='2*n_jobs', error_score=nan, return_train_score=False)
 |  
 |  Exhaustive search over specified parameter values for an estimator.
 |  
 |  Important members are fit, predict.
 |  
 |  GridSearchCV implements a "fit" and a "score" method.
 |  It also implements "score_samples", "predict", "predict_proba",
 |  "decision_function", "transform" and "inverse_transform" if they are
 |  implemented in the estimator used.
 |  
 |  The parameters of the estimator used to apply these methods are optimized
 |  by cross-validated grid-search over a parameter grid.
 |  
 |  Read more in the :ref:`User Guide <grid_search>`.
 |  
 |  Parameters
 |  ----------
 |  estimator : estimator object
 |      This is assumed to implement the scikit-learn estimator interface.
 |      Either est

NameError: name 'svc' is not defined

In [13]:
svc=SVC()
svc.fit(X_train,y_train)
params={'C':[0.01,0.1,1,10,100,1000,10000,100000]}
model_cv=GridSearchCV(svc,param_grid=params,scoring='accuracy',return_train_score=True,verbose=3)
model_cv.fit(X_train,y_train)

Fitting 5 folds for each of 8 candidates, totalling 40 fits
[CV 1/5] END ........C=0.01;, score=(train=0.675, test=0.685) total time=   0.5s
[CV 2/5] END ........C=0.01;, score=(train=0.681, test=0.652) total time=   0.5s
[CV 3/5] END ........C=0.01;, score=(train=0.677, test=0.668) total time=   0.5s
[CV 4/5] END ........C=0.01;, score=(train=0.671, test=0.693) total time=   0.5s
[CV 5/5] END ........C=0.01;, score=(train=0.672, test=0.682) total time=   0.5s
[CV 1/5] END .........C=0.1;, score=(train=0.692, test=0.696) total time=   0.4s
[CV 2/5] END .........C=0.1;, score=(train=0.698, test=0.668) total time=   0.4s
[CV 3/5] END .........C=0.1;, score=(train=0.695, test=0.680) total time=   0.4s
[CV 4/5] END .........C=0.1;, score=(train=0.689, test=0.713) total time=   0.4s
[CV 5/5] END .........C=0.1;, score=(train=0.691, test=0.713) total time=   0.4s
[CV 1/5] END ...........C=1;, score=(train=0.717, test=0.703) total time=   0.4s
[CV 2/5] END ...........C=1;, score=(train=0.708,

GridSearchCV(estimator=SVC(),
             param_grid={'C': [0.01, 0.1, 1, 10, 100, 1000, 10000, 100000]},
             return_train_score=True, scoring='accuracy', verbose=3)

In [15]:
best_alpha=model_cv.best_params_['C']

In [ ]:
best_alpha

In [12]:
y_test.shape

(1381, 1)

In [17]:
svc=SVC(C=best_alpha)
model=svc.fit(X_train,y_train)
y_pred=model.predict(X_test)


In [18]:
help(confusion_matrix)

Help on function confusion_matrix in module sklearn.metrics._classification:

confusion_matrix(y_true, y_pred, *, labels=None, sample_weight=None, normalize=None)
    Compute confusion matrix to evaluate the accuracy of a classification.
    
    By definition a confusion matrix :math:`C` is such that :math:`C_{i, j}`
    is equal to the number of observations known to be in group :math:`i` and
    predicted to be in group :math:`j`.
    
    Thus in binary classification, the count of true negatives is
    :math:`C_{0,0}`, false negatives is :math:`C_{1,0}`, true positives is
    :math:`C_{1,1}` and false positives is :math:`C_{0,1}`.
    
    Read more in the :ref:`User Guide <confusion_matrix>`.
    
    Parameters
    ----------
    y_true : array-like of shape (n_samples,)
        Ground truth (correct) target values.
    
    y_pred : array-like of shape (n_samples,)
        Estimated targets as returned by a classifier.
    
    labels : array-like of shape (n_classes), default=

In [19]:
confusion=confusion_matrix(y_true=y_test,y_pred=y_pred)

In [20]:
confusion

array([[776,  35],
       [ 59, 511]])